api/src/main/java/org/apache/iceberg/stats/StatsUtil.java

```diff
@@ -0,0 +1,223 @@
+            "value_count",
+            Types.LongType.get(),
+            "Total value count, including null and NaN"),
+        optional(
+            id + NULL_VALUE_COUNT_OFFSET,
+            "null_value_count",
+            Types.LongType.get(),
+            "Total null value count"),
+        optional(
+            id + NAN_VALUE_COUNT_OFFSET,
+            "nan_value_count",
+            Types.LongType.get(),
+            "Total NaN value count"),
+        optional(
+            id + AVG_VALUE_SIZE_OFFSET,
+            "avg_value_size",
+            Types.IntegerType.get(),
+            "Avg value size of variable-length types (String, Binary)"),
+        optional(
+            id + MAX_VALUE_SIZE_OFFSET,
+            "max_value_size",
+            Types.IntegerType.get(),
+            "Max value size of variable-length types (String, Binary)"),
+        optional(id + LOWER_BOUND_OFFSET, "lower_bound", type, "Lower bound"),
+        optional(id + UPPER_BOUND_OFFSET, "upper_bound", type, "Upper bound"));
+  }
+
+  private static class ContentStatsSchemaVisitor extends TypeUtil.SchemaVisitor<Types.NestedField> {
+    private final List<Types.NestedField> statsFields = Lists.newArrayList();
+    private final Set<Integer> skippedFieldIds = Sets.newLinkedHashSet();
+
+    @Override
+    public Types.NestedField schema(Schema schema, Types.NestedField structResult) {
+      return optional(
+          146,
+          "content_stats",
+          Types.StructType.of(
+              statsFields.stream()
+                  .filter(Objects::nonNull)
+                  .sorted(Comparator.comparing(Types.NestedField::fieldId))
+                  .collect(Collectors.toList())));
+    }
+
+    @Override
+    public Types.NestedField list(Types.ListType list, Types.NestedField elementResult) {
+      list.fields()
+          .forEach(
+              field -> {
+                Types.NestedField result = field(field, null);
+                if (null != result) {
+                  statsFields.add(result);
+                }
+              });
+      return null;
+    }
+
+    @Override
+    public Types.NestedField map(
+        Types.MapType map, Types.NestedField keyResult, Types.NestedField valueResult) {
+      map.fields()
+          .forEach(
+              field -> {
+                Types.NestedField result = field(field, null);
+                if (null != result) {
+                  statsFields.add(result);
+                }
+              });
+      return null;
+    }
+
+    @Override
+    public Types.NestedField struct(Types.StructType struct, List<Types.NestedField> fields) {
+      statsFields.addAll(fields);
+      return null;
+    }
+
+    @Override
+    public Types.NestedField field(Types.NestedField field, Types.NestedField fieldResult) {
+      if (field.type().isNestedType() || field.type().isVariantType()) {
+        return null;
+      }
+
+      int fieldId = StatsUtil.statsFieldIdForField(field.fieldId());
```

Comment:
nit: fieldId -> statsFieldStartingId. also don't need `StatsUtils.` since it is in the same class.